src/modules/ods/ods.service.ts

```diff
@@ -0,0 +1,105 @@
+   * @throws {InternalServerErrorException} If there is an error trying to find a customer
+   * @throws {NotFoundException} If no matching customer is found
+   */
+  async findCustomer(urn: string): Promise<GetOdsCustomerResponse> {
+    try {
+      const spInput = this.createOdsStoredProcedureInput(ODS_ENTITIES.CUSTOMER, { customer_party_unique_reference_number: urn });
+
+      const storedProcedureResult = await this.callOdsStoredProcedure(spInput);
+
+      const storedProcedureJson: OdsStoredProcedureOuputBody = JSON.parse(storedProcedureResult[0]?.output_body);
+
+      if (storedProcedureJson === undefined || storedProcedureJson?.status != 'SUCCESS') {
+        this.logger.error('Error from ODS stored procedure, output: %o', storedProcedureResult);
+        throw new InternalServerErrorException('Error trying to find a customer');
+      }
+
+      if (storedProcedureJson?.total_result_count === 0) {
+        throw new NotFoundException('No matching customer found');
+      }
+
+      return {
+        urn: storedProcedureJson.results[0]?.customer_party_unique_reference_number,
+        name: storedProcedureJson.results[0]?.customer_name,
+      };
+    } catch (err) {
+      if (err instanceof NotFoundException) {
+        this.logger.warn(err);
+        throw err;
+      } else {
+        this.logger.error(err);
+        throw new InternalServerErrorException('Error trying to find a customer');
+      }
+    }
+  }
+
+  /**
+   * Creates the input parameter for the stored procedure
+   * @param {OdsEntity} entityToQuery The entity you want to query in ODS
+   * @param {OdsStoredProcedureQueryParams} queryParameters The query parameters and filters to apply to the query
+   *
+   * @returns {OdsStoredProcedureInput} The ODS stored procedure input in object format
+   */
+  createOdsStoredProcedureInput(entityToQuery: OdsEntity, queryParameters: OdsStoredProcedureQueryParams): OdsStoredProcedureInput {
+    return {
+      query_method: 'get',
+      query_object: entityToQuery,
+      query_page_size: 1,
+      query_page_index: 1,
+      query_parameters: queryParameters,
+    };
+  }
+
+  /**
+   * Calls the ODS stored procedure with the input provided and returns the output of it
+   * @param {OdsStoredProcedureInput} storedProcedureInput The input parameter of the stored procedure
+   *
+   * @returns {Promise<OdsStoredProcedureOuput>} The result of the stored procedure
+   */
+  async callOdsStoredProcedure(storedProcedureInput: OdsStoredProcedureInput): Promise<OdsStoredProcedureOutput> {
+    const queryRunner = this.odsDataSource.createQueryRunner();
+    try {
+      // Use the query runner to call a stored procedure
+      const result = await queryRunner.query(
+        `
+        DECLARE @output_body NVARCHAR(MAX);
+        EXEC t_apim.sp_ODS_query @input_body=@0, @output_body=@output_body OUTPUT;
+        SELECT @output_body as output_body
+      `,
+        [JSON.stringify(storedProcedureInput)],
```

Comment:
* Is `storedProcedureInput` validated before passing it to the query?
* Does SP has enought validation to ensure no malcious code wil be executed.
* Does SP has correct ACL i.e. role management?
* Error handling should also be sanitized to ensure error messages do not expose anything sensitive back to the client.

Reply:
- we create storedProcedureInput from code and from a validated input to the endpoint (urn).
- we use parameterisation to stop SQL injection
- The user (service principal) calling this stored procedure has a very tight ACL, it can only run this stored procedure and query one logs table, it can't do anything else
- None of our error handling uses the output of the stored procedure, we only log that to Azure server logs